File: main-pl.js
const { firefox } = require('playwright');
const fs = require('fs');
const path = require('path');
const axios = require('axios');
const { parse } = require('csv-parse/sync');

async function fetchKeywords(inputCsvPath, inputKeywords) {
    let keywords = [];
    if (inputCsvPath && fs.existsSync(inputCsvPath)) {
        const fileContent = fs.readFileSync(inputCsvPath, 'utf-8');
        const records = parse(fileContent, { columns: false, skip_empty_lines: true });
        keywords = records.flat();
    }
    if (inputKeywords) {
        keywords = [...keywords, ...inputKeywords.split(',').map(k => k.trim())];
    }
    return keywords.filter(Boolean);
}

async function searchKeyword(page, keyword, searchType) {
    try {
        const searchQuery = searchType === 'intitle' 
            ? `intitle:"${keyword}"`
            : `allintitle:"${keyword}"`;

        await page.goto(`https://www.google.com/search?q=${encodeURIComponent(searchQuery)}`);
        
        const statsElement = await page.waitForSelector('#result-stats', { 
            timeout: 20000,
            state: 'attached'
        });

        if (!statsElement) {
            console.log(`No stats found for keyword: ${keyword} (${searchType})`);
            return { keyword, searchType, count: 0 };
        }

        const resultStats = await statsElement.textContent();
        const match = resultStats.match(/About ([\d,]+) results/);
        const count = match ? parseInt(match[1].replace(/,/g, ''), 10) : 0;
        
        console.log(`Keyword: "${keyword}", Type: "${searchType}", Count: ${count}`);
        return { keyword, searchType, count };

    } catch (error) {
        console.error(`Error for "${keyword}" (${searchType}): ${error.message}`);
        return null;
    }
}

async function startCrawler(keywords, id) {
    const results = [];
    const retryList = [];

    const browser = await firefox.launch({ 
        headless: true,
        args: ['--no-sandbox']
    });
    
    const page = await browser.newPage();
    page.setDefaultTimeout(3000);

    // First pass: Process all keywords with both search types
    for (const keyword of keywords) {
        // Process intitle search
        const intitleResult = await searchKeyword(page, keyword, 'intitle');
        if (intitleResult) {
            results.push(intitleResult);
        } else {
            retryList.push({ keyword, searchType: 'intitle' });
        }

        // Add delay between searches
        await page.waitForTimeout(2000);

        // Process allintitle search
        const allintitleResult = await searchKeyword(page, keyword, 'allintitle');
        if (allintitleResult) {
            results.push(allintitleResult);
        } else {
            retryList.push({ keyword, searchType: 'allintitle' });
        }

        // Add delay between keywords
        await page.waitForTimeout(2000);
    }

    // Retry failed requests
    if (retryList.length > 0) {
        console.log(`Retrying ${retryList.length} failed searches...`);
        for (const { keyword, searchType } of retryList) {
            await page.waitForTimeout(5000); // Longer delay for retries
            const retryResult = await searchKeyword(page, keyword, searchType);
            results.push(retryResult || { 
                keyword, 
                searchType, 
                count: 0 
            });
        }
    }

    await browser.close();

    // Ensure results directory exists
    const resultsDir = path.join(process.cwd(), 'results');
    fs.mkdirSync(resultsDir, { recursive: true });

    // Save results as CSV
    const resultPath = path.join(resultsDir, `${id}.csv`);
    const csvContent = 'Keyword,Search Type,Count\n' +
        results
            .sort((a, b) => a.keyword.localeCompare(b.keyword) || a.searchType.localeCompare(b.searchType))
            .map(r => `${r.keyword},${r.searchType},${r.count}`)
            .join('\n');
    
    fs.writeFileSync(resultPath, csvContent, 'utf-8');
    console.log(`Results saved to ${resultPath}`);
    
    return results;
}

async function main() {
    const [,, id, inputKeywords, inputCsvPath] = process.argv;
    
    if (!id) {
        console.error('Error: ID is required to save the results.');
        process.exit(1);
    }

    console.log(`Starting crawler with ID: ${id}`);
    console.log(`Input Keywords: ${inputKeywords}`);
    console.log(`Input CSV Path: ${inputCsvPath}`);

    try {
        const keywords = await fetchKeywords(inputCsvPath, inputKeywords);
        if (keywords.length === 0) {
            console.error('No keywords provided. Exiting...');
            process.exit(1);
        }

        console.log(`Processing ${keywords.length} keywords: ${keywords.join(', ')}`);
        const results = await startCrawler(keywords, id);
        
        const totalKeywords = keywords.length;
        const totalSearches = results.length;
        const successfulSearches = results.filter(r => r.count > 0).length;
        
        console.log(`
Processing Summary:
------------------
Total Keywords: ${totalKeywords}
Total Searches: ${totalSearches}
Successful Searches: ${successfulSearches}
Failed Searches: ${totalSearches - successfulSearches}
        `);
        
        process.exit(0);
    } catch (error) {
        console.error('Fatal error:', error);
        process.exit(1);
    }
}

process.on('uncaughtException', (error) => {
    console.error('Uncaught Exception:', error);
    process.exit(1);
});

process.on('unhandledRejection', (error) => {
    console.error('Unhandled Rejection:', error);
    process.exit(1);
});

main();

File: main-pl.py
from playwright.sync_api import sync_playwright
import csv
import os
import sys
from typing import List, Dict
import re

def fetch_keywords(input_csv_path: str = None, input_keywords: str = None) -> List[str]:
    """Fetch keywords from either a CSV file or from the input string."""
    keywords = []
    
    # If a CSV file path is provided, parse the CSV
    if input_csv_path and os.path.exists(input_csv_path):
        with open(input_csv_path, 'r', encoding='utf-8') as file:
            reader = csv.reader(file)
            keywords.extend([item for sublist in reader for item in sublist])
    
    # If keywords are provided as a string, split by commas
    if input_keywords:
        keywords.extend([k.strip() for k in input_keywords.split(',')])
    
    # Remove empty strings and duplicates
    return [k for k in keywords if k]

def start_crawler(keywords: List[str], id: str) -> None:
    """Start the crawler and process keywords."""
    results = []
    retry_list = []
    
    with sync_playwright() as p:
        browser = p.firefox.launch(headless=True)
        page = browser.new_page()
        
        # Loop through keywords and start scraping
        for keyword in keywords:
            try:
                # Open search URL
                page.goto(f'https://www.google.com/search?q=intitle%3A%22{keyword}%22')
                
                # Wait for Google's result stats
                visible = page.locator('#result-stats').is_visible()
                if not visible:
                    print('There is no stats showing at all')
                    continue
                
                t = page.locator('#result-stats').all_text_contents()
                print('========', t)
                
                # Extract result count
                result_stats = page.locator('#result-stats').text_content()
                match = re.search(r'About ([\d,]+) results', result_stats)
                count = int(match.group(1).replace(',', '')) if match else 0
                
                results.append({
                    'keyword': keyword,
                    'searchType': 'intitle',
                    'count': count
                })
                print(f'Keyword: "{keyword}", Type: "intitle", Count: {count}')
                
            except Exception as error:
                print(f'Error for "{keyword}": {str(error)}')
                retry_list.append(keyword)
        
        # Retry failed requests
        if retry_list:
            print(f'Retrying {len(retry_list)} failed keywords...')
            for keyword in retry_list:
                try:
                    page.goto(f'https://www.google.com/search?q=intitle%3A%22{keyword}%22')
                    page.wait_for_selector('#result-stats', timeout=10000)
                    result_stats = page.locator('#result-stats').text_content()
                    match = re.search(r'About ([\d,]+) results', result_stats)
                    count = int(match.group(1).replace(',', '')) if match else 0
                    
                    results.append({
                        'keyword': keyword,
                        'searchType': 'intitle',
                        'count': count
                    })
                    print(f'Retrying Keyword: "{keyword}", Type: "intitle", Count: {count}')
                    
                except Exception as error:
                    print(f'Error retrying "{keyword}": {str(error)}')
        
        browser.close()
    
    # Save results as CSV
    result_path = os.path.join(os.path.dirname(__file__), 'results', f'{id}.csv')
    os.makedirs(os.path.dirname(result_path), exist_ok=True)
    
    with open(result_path, 'w', newline='', encoding='utf-8') as file:
        writer = csv.DictWriter(file, fieldnames=['keyword', 'searchType', 'count'])
        writer.writeheader()
        writer.writerows(results)
    
    print(f'Results saved to {result_path}')

def main():
    """Main function to run the crawler."""
    # Get command line arguments
    args = sys.argv[1:]
    if len(args) < 1:
        print('Error: ID is required to save the results.')
        sys.exit(1)
    
    id = args[0]
    input_keywords = args[1] if len(args) > 1 else None
    input_csv_path = args[2] if len(args) > 2 else None
    
    print(f'Starting crawler with ID: {id}')
    
    keywords = fetch_keywords(input_csv_path, input_keywords)
    if not keywords:
        print('No keywords provided. Exiting...')
        sys.exit(1)
    
    print(f'Fetched Keywords: {", ".join(keywords)}')
    start_crawler(keywords, id)

if __name__ == '__main__':
    main()
